hashes/zkevm/src/keccak/coprocessor/circuit/leaf.rs

```diff
@@ -0,0 +1,501 @@
+                let hash_hi = copy_manager.load_external_assigned(output_row.hash_hi);
+                LoadedKeccakF { bytes_left, word_values, is_final, hash_lo, hash_hi }
+            })
+            .collect()
+    }
+
+    /// Generate witnesses of the base circuit.
+    fn generate_base_circuit_witnesses(&self, loaded_keccak_fs: &[LoadedKeccakF<F>]) {
+        let range = self.base_circuit_builder.borrow().range_chip();
+        let gate = range.gate();
+        let circuit_final_outputs = {
+            let mut base_circuit_builder_mut = self.base_circuit_builder.borrow_mut();
+            let ctx = base_circuit_builder_mut.main(0);
+            let mut hasher = self.hasher.borrow_mut();
+            hasher.initialize_consts(ctx, gate);
+
+            let lookup_key_per_keccak_f =
+                encode_inputs_from_keccak_fs(ctx, gate, &hasher, loaded_keccak_fs);
+            Self::generate_circuit_final_outputs(
+                ctx,
+                gate,
+                &lookup_key_per_keccak_f,
+                loaded_keccak_fs,
+            )
+        };
+        self.publish_outputs(&circuit_final_outputs);
+    }
+
+    /// Combine lookup keys and Keccak results to generate final outputs of the circuit.
+    fn generate_circuit_final_outputs(
+        ctx: &mut Context<F>,
+        gate: &impl GateInstructions<F>,
+        lookup_key_per_keccak_f: &[PoseidonCompactOutput<F>],
+        loaded_keccak_fs: &[LoadedKeccakF<F>],
+    ) -> Vec<KeccakCircuitOutput<AssignedValue<F>>> {
+        let KeccakCircuitOutput {
+            key: dummy_key_val,
+            hash_lo: dummy_keccak_val_lo,
+            hash_hi: dummy_keccak_val_hi,
+        } = dummy_circuit_output::<F>();
+
+        // Dummy row for keccak_fs with is_final = false. The corresponding logical input is empty.
+        let dummy_key_witness = ctx.load_constant(dummy_key_val);
+        let dummy_keccak_lo_witness = ctx.load_constant(dummy_keccak_val_lo);
+        let dummy_keccak_hi_witness = ctx.load_constant(dummy_keccak_val_hi);
+
+        let mut circuit_final_outputs = Vec::with_capacity(loaded_keccak_fs.len());
+        for (compact_output, loaded_keccak_f) in
+            lookup_key_per_keccak_f.iter().zip(loaded_keccak_fs)
+        {
+            let is_final = AssignedValue::from(loaded_keccak_f.is_final);
+            let key = gate.select(ctx, *compact_output.hash(), dummy_key_witness, is_final);
+            let hash_lo =
+                gate.select(ctx, loaded_keccak_f.hash_lo, dummy_keccak_lo_witness, is_final);
+            let hash_hi =
+                gate.select(ctx, loaded_keccak_f.hash_hi, dummy_keccak_hi_witness, is_final);
+            circuit_final_outputs.push(KeccakCircuitOutput { key, hash_lo, hash_hi });
+        }
+        circuit_final_outputs
+    }
+
+    /// Publish outputs of the circuit as public instances.
+    fn publish_outputs(&self, outputs: &[KeccakCircuitOutput<AssignedValue<F>>]) {
```

Comment:
@jonathanpwang because `base_circuit_builder` will be borrowed in this function, we intend to not pass `range`/`gate` which comes from `base_circuit_builder`.

Reply:
ah, `base_circuit_builder.range_chip()` creates an entirely new `RangeChip`. This was necessary just to ensure it was sharing the correct `LookupAnyManager`.

anyways all a new range chip does is that it recomputes some powers of 2 constants and clones smart pointers. So either way is fine.

Reply:
I want to make sure `base_circuit_builder` is released when entering a new function because I ran into some borrowing issues. I think we probably should not take the assumption `RangeChip` doesn't keep any reference about `base_circuit_builder`.